tests/Unit/Service/Client/ServiceClientTest.php

```diff
@@ -213,6 +219,68 @@ public function testItCanPerformAServiceCallFromPopulatedTokenCacheWithScopes():
         $this->assertEquals('service response', $result->getBody()->__toString());
     }
 
+    public function testItCanPerformAServiceCallFromPopulatedInvalidTokenCacheAndAutoRetry(): void
+    {
+        $scopes = ['scope1', 'scope2'];
+
+        $cacheKey = $this->generateAccessTokenCacheKey($this->registration, $scopes);
+        $cacheItem = $this->cache->getItem($cacheKey)->set('invalid_access_token');
+        $this->cache->save($cacheItem);
+
+        $this->clientMock
+            ->expects($this->exactly(3))
+            ->method('request')
+            ->withConsecutive(
+                [
+                    'GET',
+                    'http://example.com',
+                    [
+                        'headers' => ['Authorization' => 'Bearer invalid_access_token'],
+                        'http_errors' => true
+                    ]
+                ],
+                [
+                    'POST',
+                    $this->registration->getPlatform()->getOAuth2AccessTokenUrl(),
+                    [
+                        'form_params' => [
+                            'grant_type' => ServiceClientInterface::GRANT_TYPE,
+                            'client_assertion_type' => ClientAssertionCredentialsGrant::CLIENT_ASSERTION_TYPE,
+                            'client_assertion' => $this->createTestClientAssertion($this->registration),
+                            'scope' => 'scope1 scope2'
+                        ]
+                    ]
+                ],
+                [
+                    'GET',
+                    'http://example.com',
+                    [
+                        'headers' => ['Authorization' => 'Bearer valid_access_token'],
+                        'http_errors' => true
+                    ]
+                ]
+            )
+            ->willReturnOnConsecutiveCalls(
+                $this->throwException(
+                    new ClientException(
+                        'invalid token',
+                        $this->createMock(ServerRequestInterface::class),
+                        $this->createResponse('invalid token', 401)
+                    )
+                ),
+                $this->createResponse(json_encode(['access_token'=> 'valid_access_token', 'expires_in' => 3600]), 201),
+                $this->createResponse('service response')
```

Comment:
No biggie, but I guess one more test case can be added when
1. First request fails with a 401
2. Access token is requested
3. Following re-request fails for some other reason

In such a scenario, we have to be sure to throw an LTI exception too.

Reply:
added this here: https://github.com/oat-sa/lib-lti1p3-core/pull/86/commits/5c867274864a4507f203c4bf66cc9d040f95e5c9